<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>rRIF: An R implementation of the Regulatory Impact Factors (RIF) • rRIF</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="rRIF: An R implementation of the Regulatory Impact Factors (RIF)">
<meta property="og:description" content="rRIF">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rRIF</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">rRIF: An R implementation of the Regulatory Impact Factors (RIF)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>rRIF: An R implementation of the Regulatory Impact Factors (RIF)</h1>
                        <h4 data-toc-skip class="author">Atsushi Fukushima</h4>
            <address class="author_afil">
      RIKEN Center for Sustainable Resource Science<br><a class="author_email" href="mailto:#"></a><a href="mailto:afukushima@gmail.com" class="email">afukushima@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2021-04-29</h4>
      
      
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor" aria-hidden="true"></a>Introduction</h1>
<p>Regulatory Impact factors (RIF) [1-2] is to identify critical transcriptional regulators like transcription factors (TFs) from gene expression data based on microarrays and RNA-seq.</p>
<p>The rRIF package is an R implementation of the RIF presented by Hudson et al.  [1] and Reverter et al. [2]. RIF aims to identify critical transcription factor (TF) between two conditions from gene expression data. The concept is based on integration of three statistical information from typical transcriptome data: (1) the co-expression change between the TF and the differentially expressed genes (DEGs); (2) the amount of differential expression (i.e, foldchange) of DEGs; and (3) the abundance of DEGs. The package can calculate both RIF1 and RIF2 and expand the original HudsonRIF’s code.</p>
<p>This vignette is largely based on the paper by Bottje et al. [3]. Basically, you can calculate RIF by using a list of DEGs and regulator information of your intereset. You should use appropriate gene identifiers or probeset-IDs.</p>
</div>
<div id="methods---the-mathematical-formalism" class="section level1">
<h1 class="hasAnchor">
<a href="#methods---the-mathematical-formalism" class="anchor" aria-hidden="true"></a>Methods - the mathematical formalism</h1>
<div id="expression" class="section level2">
<h2 class="hasAnchor">
<a href="#expression" class="anchor" aria-hidden="true"></a>Expression</h2>
<p>Average (normalized) expression of the <span class="math inline">\(i\)</span>-th gene across the <span class="math inline">\(n\)</span> samples (eg. biological replicates) of experimental condition A and where each <span class="math inline">\(x_{i,k}\)</span> corresponds to the expression of the <span class="math inline">\(i\)</span>-th gene in the <span class="math inline">\(k\)</span>-th sample (<span class="math inline">\(k\)</span> = 1, …, <span class="math inline">\(n\)</span>).</p>
<p><span class="math display">\[ E_{i,A} = \frac {1}{n} \sum_{k=1}^{n} x_{i,k} \]</span></p>
</div>
<div id="differential-expression-de" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-de" class="anchor" aria-hidden="true"></a>Differential expression (DE)</h2>
<p>Difference in the expression of the <span class="math inline">\(i\)</span>-th gene in the two conditions under scrutiny, A and B (eg. healthy and diseased, two breeds, two diets, two time points, …). Note that it is not a requirement to have the same number of samples surveyed in the two conditions.</p>
<p><span class="math display">\[ dE_{i} = E_{i,A} - E_{i,B}\]</span></p>
</div>
<div id="co-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#co-expression" class="anchor" aria-hidden="true"></a>Co-expression</h2>
<p>Similarity of expression profile (typically and shown here the Spearman correlation coefficient) between the <span class="math inline">\(i\)</span>-th and the <span class="math inline">\(j\)</span>-th genes across the <span class="math inline">\(n\)</span> samples of condition A.</p>
<p><span class="math display">\[ C_{i,j} = r_A(i,j) = \frac {Cov(i,j)}{\sigma_i \sigma_j} \]</span> ## Diffenrential co-expression or differential wiring in co-expression networks Difference in the co-expression between the <span class="math inline">\(i\)</span>-th and the <span class="math inline">\(j\)</span>-th genes in the two conditions under scrutiny, A and B. Note that it is not a requirement to have the same number of samples surveyed in the two conditions.</p>
<p><span class="math display">\[ dC_{i,j} = r_A(i,j) - r_B(i,j)\]</span></p>
</div>
<div id="co-differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#co-differential-expression" class="anchor" aria-hidden="true"></a>Co-differential expression</h2>
<p>Similarity of the profile of differential expression of genes <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> across the levels of another experimental design effect such as time points. Two conditions, A and B, are being surveyed across a series of developmental time points.</p>
<p><span class="math display">\[ CdE_{i,j} = r(dE_i, dE_j) \]</span></p>
</div>
<div id="phenotype-impact-factor-pif" class="section level2">
<h2 class="hasAnchor">
<a href="#phenotype-impact-factor-pif" class="anchor" aria-hidden="true"></a>Phenotype Impact Factor (PIF)</h2>
<p>Average (normalized) expression of the <span class="math inline">\(i\)</span>-th gene across the two conditions multiplied by its differential expression. In other words, PIF weights the differential expression of a given gene by its overall abundance.</p>
<p><span class="math display">\[ PIF_i = \frac {1}{2} (E_{i,A} + E_{i,B})dE_i = A_idE_i \]</span></p>
</div>
<div id="regulatory-impact-factor-option-1" class="section level2">
<h2 class="hasAnchor">
<a href="#regulatory-impact-factor-option-1" class="anchor" aria-hidden="true"></a>Regulatory Impact Factor, Option 1</h2>
<p>For the <span class="math inline">\(i\)</span>-th regulator and across all the j differentially expressed genes (<span class="math inline">\(j\)</span> = 1, …, <span class="math inline">\(n_{dE}\)</span>) RIF1 looks at the average PIF of the <span class="math inline">\(i\)</span>-th regulator weighted by the squared differential co-expression between the <span class="math inline">\(i\)</span>-th regulator and the <span class="math inline">\(j\)</span>-th differentially expressed gene. It addresses the question: Which regulator is consistently highly differentially co-expressed with the abundant differentially expressed gene?</p>
<p><span class="math display">\[ RIF1_i = \frac {1}{n_{dE}} \sum_{j=1}^{j=n_{dE}} (PIF_j dC_{i,j}^2) \]</span></p>
</div>
<div id="regulatory-impact-factor-option-2" class="section level2">
<h2 class="hasAnchor">
<a href="#regulatory-impact-factor-option-2" class="anchor" aria-hidden="true"></a>Regulatory Impact Factor, Option 2</h2>
<p>For the <span class="math inline">\(i\)</span>-th regulator and across all the <span class="math inline">\(j\)</span> differentially expressed genes (<span class="math inline">\(j\)</span> = 1, …, <span class="math inline">\(n_{dE}\)</span> ) RIF1 looks at the average change in predictive ability of the <span class="math inline">\(i\)</span>-th regulator to predict the abundance of the <span class="math inline">\(j\)</span>-th differentially expressed gene. It addresses the question: Which regulator has the most altered ability to predict the abundance of differentially expressed genes.</p>
<p><span class="math display">\[ RIF2_i = \frac {1}{n_{dE}} [(E_{j,A} r_A(i,j))^2 - (E_{j,B} r_B(i,j))^2] \]</span></p>
</div>
</div>
<div id="prerequisites" class="section level1">
<h1 class="hasAnchor">
<a href="#prerequisites" class="anchor" aria-hidden="true"></a>Prerequisites</h1>
<p>The rRIF package requires (1) a list of DEGs and (2) a list of regulators of interest. For example, we can use the transription factor database like <a href="http://planttfdb.cbi.pku.edu.cn/" class="external-link">PlantTFDB</a>.</p>
<p>The rRIF package is organisms-free. It means that you can do that based on your DEGs list and pre-defined regulator list. You must use the same gene- or probeset IDs in the analysis.</p>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor" aria-hidden="true"></a>Getting started</h1>
<div id="datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#datasets" class="anchor" aria-hidden="true"></a>Datasets</h2>
<p>In this vignette, we use example datasets presented by [Bottje et al. [3]. This is a transcriptome data matrix. The data are composed of 520 significant DEGs (differentially expressed genes) and 898 transcription factors.</p>
</div>
<div id="rif" class="section level2">
<h2 class="hasAnchor">
<a href="#rif" class="anchor" aria-hidden="true"></a>RIF</h2>
<p>Using these example datasets, we can perform the RIF as follows.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rRIF.html">rRIF</a></span><span class="op">(</span>eSet <span class="op">=</span> <span class="va">ToniData</span>,
            formula <span class="op">=</span> <span class="va">geno1</span><span class="op">~</span><span class="va">geno2</span>,
            target.factor <span class="op">=</span> <span class="st">"Genotype"</span>,
            DEGs <span class="op">=</span> <span class="va">ToniData.DEGs</span>,
            cor.method <span class="op">=</span> <span class="st">"pearson"</span>,
            regulator.list <span class="op">=</span> <span class="va">ToniData.TFs</span>
            <span class="op">)</span></code></pre></div>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## 
## ===== rRIF Summary =====
## Top 5 RIF1:</code></pre>
<pre><code>##              A    DE   PIF  RIF1  RIF2
## CBFA2T3_1 6.80  0.12  0.83 -5.56 -0.69
## SS18_1    9.43 -0.10 -0.98 -4.93 -0.87
## HIC1_1    8.51  0.01  0.09 -3.95 -0.54
## TAF1A_1   7.47  0.07  0.51 -3.81 -0.51
## MAFK_1    7.23  0.32  2.29 -3.65 -0.19</code></pre>
<p><code>rRIF</code> function() calculates both RIF1 and RIF2 and stores their results. Each RIF value was z-transformed by <code>scale</code> function() according to the original papers regarding RIF (Hudson et al. 2009; Reverter et al. 2010). You should get the integrated results like this.</p>
<pre><code>##     AATF_1   ACTL6A_1    ACTR5_1    ACTR8_1     ADD1_1 
## -0.2263297  0.2569174 -2.0209675  0.6461308 -1.1604140</code></pre>
<p>You can also calculate the RIF1 and RIF2 separately using <code>calculateRIF1</code> and <code>calculateRIF2</code>.</p>
</div>
<div id="pif" class="section level2">
<h2 class="hasAnchor">
<a href="#pif" class="anchor" aria-hidden="true"></a>PIF</h2>
<p>We can perform the PIF as follows.</p>
<pre><code>##  5-MCDG_1    AATF_1  ABI3BP_1  ABLIM3_1   ACACB_1     ACE_1 
## -1.182956  0.860097 -8.097313  1.843308 -5.319971 -1.879602</code></pre>
</div>
</div>
<div id="acknowledgments" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgments" class="anchor" aria-hidden="true"></a>Acknowledgments</h1>
<p>We thank Toni (Antonio Reverter) and Nick (Nicholas J. Hudson) for helpful discussions and for providing us with original data published by Bottje et al.  (2017). This work was supported by JSPS KAKENHI Grant Numbers 26850024 and 17K07663.</p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor" aria-hidden="true"></a>References</h1>
<ol style="list-style-type: decimal">
<li>Hudson NJ et al. “A differential wiring analysis of expression data correctly identifies the gene containing the causal mutation”, PLoS Comput Biol 5:e1000382 (2009) <a href="https://www.ncbi.nlm.nih.gov/pubmed/19412532" class="external-link">PubMed</a>
</li>
<li>Reverter A et al. “Regulatory impact factors: unraveling the transcriptional regulation of complex traits from expression data”, Bioinformatics 26:896-904 (2010) <a href="https://www.ncbi.nlm.nih.gov/pubmed/20144946" class="external-link">PubMed</a>
</li>
<li>Bottje W et al. “Progesterone signalling in broiler skeletal muscle is associated with divergent feed efficiency”, BMC Systems Biology 11:29 (2017) <a href="https://www.ncbi.nlm.nih.gov/pubmed/28235404" class="external-link">PubMed</a>
</li>
</ol>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor" aria-hidden="true"></a>Session info</h1>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on which this document was compiled:</p>
<pre><code>## R Under development (unstable) (2021-04-15 r80175)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] Biobase_2.51.0      BiocGenerics_0.37.4 rRIF_0.99.1        
## [4] knitr_1.33         
## 
## loaded via a namespace (and not attached):
##  [1] magrittr_2.0.1     R6_2.5.0           ragg_1.1.2         rlang_0.4.10      
##  [5] fastmap_1.1.0      stringr_1.4.0      tools_4.1.0        xfun_0.22         
##  [9] jquerylib_0.1.4    htmltools_0.5.1.1  systemfonts_1.0.1  yaml_2.2.1        
## [13] digest_0.6.27      rprojroot_2.0.2    pkgdown_1.6.1.9001 crayon_1.4.1      
## [17] textshaping_0.3.3  sass_0.3.1         prettydoc_0.4.1    fs_1.5.0          
## [21] memoise_2.0.0      cachem_1.0.4       evaluate_0.14      rmarkdown_2.7     
## [25] stringi_1.5.3      compiler_4.1.0     bslib_0.2.4        desc_1.3.0        
## [29] jsonlite_1.7.2</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Atsushi Fukushima.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
